### PR TITLE
Fix OpenACC MDRangePolicy errors (issue #7403)

### PR DESCRIPTION
**Pull request by seyonglee**
This PR fixes the OpenACC MDRangePolicy errors (issue #7403).
OpenACC parallel constructs with MDRangePolicy and the new BasicView suffer from dangling pointer problems when compiled by NVHPC compiler (tested with V23.9 and 24.5). This problem disappears if loop-collapsing transformation is manually applied to the MDRangePolicy loops. This PR applies the manual loop-collapsing transformation only when compiled by NVHPC compiler.

### CODE REVIEW

**Comment by dalg24** [comment]
Any idea what that "dangling pointer" issue is about?

**Comment by seyonglee** [comment]
> Any idea what that "dangling pointer" issue is about?

When an OpenACC parallel construct with MDRangePolicy and the new BasicView is compiled by the NVHPC compiler, the execution fails with illegal address errors (`CUDA_ERROR_ILLEGAL_ADDRESS`) and NVIDIA compute-sanitizer complains multiple invalid memory accesses with the following comment: "Launch failed (often invalid pointer dereference)".
I guess that there is a NVHPC compiler bug in handling OpenACC `collapse` clause when deeply nested loops have complex functor body. (Actual behavior varies depending on the optimization level (`-O2` or higher), the complexity of functor body, and loop nesting level.) However, if collapse transformation is manually applied, those problem disappears.

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp [hunk 1]
```diff
@@ -30,10 +30,23 @@ void OpenACCParallelForMDRangePolicy(OpenACCCollapse, OpenACCIterateLeft,
...
+  auto begin0 = begin[0];
+  auto end0   = end[0];
+#if defined(KOKKOS_ENABLE_OPENACC_COLLAPSE_MDRANGE_LOOPS)
+  auto dim1  = end1 - begin1;
+  auto dim0  = end0 - begin0;
+  auto nIter = dim1 * dim0;
+// clang-format off
+#pragma acc parallel loop gang vector copyin(functor) async(async_arg)
+  // clang-format on
+  for (auto m = 0; m < nIter; ++m) {
```
Comment (crtrott) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp:44]: I appreciate the sentiment of the "auto" but `m` is going to be `int` this way, and if nIter turns out to be unsigned you still get a warning.

Comment (crtrott) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp:44]:
```suggestion
  for (decltype(nIter) m = 0; m < nIter; ++m) {
```

**Review by crtrott — COMMENTED** [review]
This feels like a LOT of extra code is it worth it?

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp:44]: You need to check for other places like this.

Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp]: Updated.

Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp]: Updated.

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp [hunk 2]
```diff
@@ -30,10 +30,23 @@ void OpenACCParallelForMDRangePolicy(OpenACCCollapse, OpenACCIterateLeft,
...
-  int begin0 = begin[0];
-  int end0   = end[0];
+  auto begin1 = begin[1];
+  auto end1   = end[1];
+  auto begin0 = begin[0];
+  auto end0   = end[0];
+#if defined(KOKKOS_ENABLE_OPENACC_COLLAPSE_MDRANGE_LOOPS)
+  auto dim1  = end1 - begin1;
+  auto dim0  = end0 - begin0;
+  auto nIter = dim1 * dim0;
```
[the remark this answers is not in the document]
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp:40]: As I know, there is no easy way to detect overflow; we will have to test values (dim0, dim1, … in this case) before performing operations (multiplication in this case), which requires to check multiple cases that can cause the overflow. Please let me know if there is any easy way to check the overflow.

**Review by tcclevenger — APPROVED** [review]

**Review by tcclevenger — APPROVED** [review]
Manually compared these indices calculation from the flat index to the multi-dim version for some test bounds and did not find any errors.

**Review by crtrott — APPROVED** [review]

**Pull request merged**